brave/brave-ios
# Fix #4352: Implement Playlist Folders

## Summary of Changes
- Implements Playlist Folders
- Adds Context Menus 
- Adds CarPlay Folders


This pull request fixes #4352

## Submitter Checklist:


## Test Plan:
Verify migration:
- install 1.35, add bunch of videos to playlist
- Upgrade to 1.36, verify that videos are there in `Saved` folder

- Verify CarPlay folders works
- Verify Create, Edit (changing the folder name), Delete on folders work (Saved Folder is the only one that cannot do this).
- Verify re-arranging/re-ordering folders works
- Verify re-ordering items within a folder works
- Verify moving items from one folder to another works
- Verify empty folders show the empty state
- Verify Playlist thumbnails show
- Verify Long Press on a Folder to bring up context menu
- Verify Long Press on a Playlist Item brings up context menu
- Verify creating a folder with existing items, moves those items into the newly created folder
- Verify cannot move items when only one folder exists
- Verify cannot move item to the same folder that it's already in
- Verify all of the above on iPad
- Verify all of the above in Landscape
- Verify the above while plugged into the Car and that the CarPlay updates when deleted from iPhone.


## Screenshots:



## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
